File: code/sorting/unclassified/src/leap_year/leap_year.cpp
#include <iostream>

bool isLeapYear(int year)
{
    if (year % 400 == 0)
        return true;

    if (year % 100 == 0)
        return false;

    return year % 4 == 0;
}

void findLeapYears(int a, int b)
{
    std::cout << "The leap years between " << a << " and " << b << " are : \n";
    for (int i = a; i <= b; i++)
        if (isLeapYear(i))
            std::cout << i << "\n";
}

int main()
{
    using namespace std;
    int year1, year2;
    cout << "Enter the first year: ";
    cin >> year1;
    cout << "Enter the second year year: ";
    cin >> year2;
    if (year1 > year2)
        findLeapYears(year2, year1);
    else
        findLeapYears(year1, year2);

    return 0;
}

File: code/sorting/unclassified/src/leap_year/leap_year.rs
// Part of Cosmos by OpenGenus
// find_leap_year takes in 2 years as start and end value
// and returns a vector containing the leap years between the given years.
// [start, end); end year in not included
fn find_leap_years(start: i32, end: i32) -> Vec<i32> {
    (start..end)
        .filter(|y| (y % 4 == 0 && y % 100 != 0) || (y % 400 == 0))
        .collect()
}

#[test]
fn name() {
    assert_eq!(
        find_leap_years(2000, 2018),
        vec![2000, 2004, 2008, 2012, 2016]
    );
}
